sw/device/lib/base/status.h

```diff
@@ -0,0 +1,135 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#include <assert.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+
+#define USING_ABSL_STATUS
+#include "sw/device/lib/base/absl_status.h"
+#undef USING_ABSL_STATUS
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * We use the error category codes from absl_status.h.  We build a packed
+ * status value that identifies the source of the error (in the form of the
+ * module identifier and line number).
+ *
+ * By default, the module identifier is the first three letters of the
+ * source filename.  The identifier can be overridden (per-module) with the
+ * DECLARE_MODULE_ID macro.
+ *
+ * Our status codes are arranged as a packed bitfield, with the sign
+ * bit signifying whether the value represents a result or an error.
+ *
+ * All Ok (good) values:
+ * 32  31                                             0
+ *  +---+---------------------------------------------+
+ *  |   |                  31 bit                     |
+ *  | 0 |                  Result                     |
+ *  +---+---------------------------------------------+
+ *
+ * All Error values:
+ * 32  31      26      21      16             5       0
+ *  +---+-------+-------+-------+-------------+-------+
+ *  |   |   15 bit              | 11 bit      | 5 bit |
+ *  | 1 |   Module Identifier   | Line Number | code  |
+ *  +---+-------+-------+-------+-------------+-------+
+ *
+ * The module identifier value is interpreted as three 5-bit fields
+ * representing the characters [0x40..0x5F] (e.g. [@ABC ... _]).
+ */
+typedef union Status_ {
+  int32_t value;
+  struct {
+    uint32_t code : 5;
+    uint32_t arg : 11;
+    uint32_t module_id : 15;
+    uint32_t is_error : 1;
+  } error;
+} status_t;
+
+// Uses a GCC statement-expr to embed conrtol-flow inside an expression.
+#define TRY(expr_)                \
+  ({                              \
+    status_t status_ = expr_;     \
+    if (status_.error.is_error) { \
+      return status_;             \
+    }                             \
+    status_.value;                \
+  })
+
+// This global constant is available to all modules and is the constant zero.
+extern const uint16_t status_module_id;
+// clang-format off
+#define ASCII_5BIT(v) ( \
+    /*uppercase characters*/  (v) >= '@' && (v) <= '_' ? (v) - '@' \
+    /*lower cvt upper*/     : (v) >= '`' && (v) <= 'z' ? (v) - '`' \
+    /*else cvt underscore*/ : '_' - '@'                            \
+  )
+// clang-format on
+
+#define MAKE_MODULE_ID(a, b, c) \
+  ASCII_5BIT(a) | (ASCII_5BIT(b) << 5) | (ASCII_5BIT(c) << 10)
```

Comment:
IMHO this is the correct encoding but it's the opposite of what we have in `error.h`. We can fix that later if you agree.

Reply:
The encoding in `silicon_creator/lib/error.h` was chosen to make it easy for humans to read the ascii codes out of the hex representation of the error code.  "5-bit ascii" packed this way is simply unreadable unless you can do the bitops in your head. (Trivia: `0x0e11` is the PCI vendor ID for Compaq Computer Corporation.  That number is 5-bit packed ascii for `CPQ`, Compaq's preferred abbreviation and their stock symbol before the HP merger).